src/main/java/io/wispforest/owo/serialization/impl/json/JsonDeserializer.java

```diff
@@ -0,0 +1,291 @@
+package io.wispforest.owo.serialization.impl.json;
+
+import com.google.gson.*;
+import io.wispforest.owo.serialization.*;
+import io.wispforest.owo.serialization.impl.SerializationAttribute;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.*;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class JsonDeserializer implements SelfDescribedDeserializer<JsonElement> {
+
+    private static final Set<SerializationAttribute> ATTRIBUTES = EnumSet.allOf(SerializationAttribute.class);
+
+    protected final Deque<Supplier<JsonElement>> stack = new ArrayDeque<>();
+
+    public JsonDeserializer(JsonElement element, boolean compressed) {
+        stack.push(() -> element);
+    }
+
+    private JsonElement topElement() {
+        return stack.peek().get();
+    }
+
+    //--
+
+    public static JsonDeserializer of(JsonElement element) {
+        return new JsonDeserializer(element, false);
+    }
+
+    public static JsonDeserializer compressed(JsonElement element) {
+        return new JsonDeserializer(element, true);
+    }
+
+    @Override
+    public Set<SerializationAttribute> attributes() {
+        return ATTRIBUTES;
+    }
+
+    //--
+
+    @Override
+    public <S> void readAny(Serializer<S> visitor) {
+        this.decodeValue(visitor, this.topElement());
+    }
+
+    private <S> void decodeValue(Serializer<S> visitor, JsonElement element) {
+        if (element.isJsonNull()) {
+            visitor.writeOptional(Endec.JSON_ELEMENT, Optional.empty());
+        } else if (element instanceof JsonPrimitive primitive) {
+            if (primitive.isString()) {
+                visitor.writeString(primitive.getAsString());
+            } else if (primitive.isBoolean()) {
+                visitor.writeBoolean(primitive.getAsBoolean());
+            } else {
+                var value = primitive.getAsBigDecimal();
+
+                try {
+                    var asLong = value.longValueExact();
+
+                    if ((byte) asLong == asLong) {
+                        visitor.writeByte(element.getAsByte());
+                    } else if ((short) asLong == asLong) {
+                        visitor.writeShort(element.getAsShort());
+                    } else if ((int) asLong == asLong) {
+                        visitor.writeInt(element.getAsInt());
+                    } else {
+                        visitor.writeLong(asLong);
+                    }
+                } catch (ArithmeticException bruh /* quite cringe java moment, why use an exception for this */) {
+                    var asDouble = value.doubleValue();
+
+                    if ((float) asDouble == asDouble) {
+                        visitor.writeFloat(element.getAsFloat());
+                    } else {
+                        visitor.writeDouble(asDouble);
+                    }
+                }
```

Comment:
couldn't you just use `.getAsNumber` and then check the type of the returned value?